server/redeemer.go

```diff
@@ -0,0 +1,398 @@
+package server
+
+// generate the go bindings for github.com/livepeer/go-livepeer/net/redeemer.proto first !
+//go:generate mockgen -source github.com/livepeer/go-livepeer/net/redeemer.pb.go -destination github.com/livepeer/go-livepeer/net/redeemer_mock.pb.go -package net
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	gonet "net"
+	"net/url"
+	"sync"
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/net"
+	"github.com/livepeer/go-livepeer/pm"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+const rpcTimeout = 8 * time.Second
+
+var cleanupLoopTime = 1 * time.Hour
+
+// Redeemer is the interface for a ticket redemption gRPC service
+type Redeemer interface {
+	net.TicketRedeemerServer
+	Start(host *url.URL) error
+	Stop()
+}
+
+type redeemer struct {
+	recipient   ethcommon.Address
+	subs        sync.Map
+	eth         eth.LivepeerEthClient
+	sm          pm.SenderMonitor
+	quit        chan struct{}
+	liveSenders sync.Map // ethCommon.Address => time.Time lastAccess
+}
+
+// NewRedeemer creates a new ticket redemption service instance
+func NewRedeemer(recipient ethcommon.Address, eth eth.LivepeerEthClient, sm pm.SenderMonitor) (Redeemer, error) {
+
+	if recipient == (ethcommon.Address{}) {
+		return nil, fmt.Errorf("must provide a recipient")
+	}
+
+	if eth == nil {
+		return nil, fmt.Errorf("must provide a LivepeerEthClient")
+	}
+
+	if sm == nil {
+		return nil, fmt.Errorf("must provide a SenderMonitor")
+	}
+
+	return &redeemer{
+		recipient: recipient,
+		eth:       eth,
+		sm:        sm,
+		quit:      make(chan struct{}),
+	}, nil
+}
+
+func (r *redeemer) Start(host *url.URL) error {
+	listener, err := gonet.Listen("tcp", host.String())
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+	if err != nil {
+		return err
+	}
+	// slice of gRPC options
+	// Here we can configure things like TLS
+	opts := []grpc.ServerOption{}
+	// var s *grpc.Server
+	s := grpc.NewServer(opts...)
+	defer s.Stop()
+
+	net.RegisterTicketRedeemerServer(s, r)
+
+	go r.startCleanupLoop()
+
+	return s.Serve(listener)
+}
+
+func (r *redeemer) Stop() {
+	close(r.quit)
+}
+
+func (r *redeemer) QueueTicket(ctx context.Context, ticket *net.Ticket) (*empty.Empty, error) {
+	t := pmTicket(ticket)
+	if err := r.sm.QueueTicket(t); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	glog.Infof("ticket queued sender=0x%x", ticket.Sender)
+
+	go r.monitorMaxFloat(ethcommon.BytesToAddress(ticket.Sender))
+	return &empty.Empty{}, nil
+}
+
+func (r *redeemer) monitorMaxFloat(sender ethcommon.Address) {
```

Comment:
If this is only called in `QueueTicket()`, then the server will only subscribe to max float updates for a sender via the SenderMonitor when a ticket is queued, but not when a sender's max float changes due to other reserve changes. We should subscribe to max float updates for a sender via the SenderMonitor as soon as a client indicates that it cares about float updates for the sender.

Reply:
Ah yes you're right ! 

But changing where this method is called wouldn't change much I think. 
This method subscribes to max float updates from the `SenderMonitor` through a channel to then send it to the RPC clients. We do this upon receiving a ticket if we are not already monitoring this sender. 

The issue is that we only receive updates on the `sink` channel from `addFloat` and `subFloat` but not when we receive ETH events. 

I think the solution would be to send updates into the subscription from `SenderMonitor.SubscribeMaxFloat`. 

Because this state is actually kept on the `SenderManager` I think we would have to create a `SubscribeMaxFloatUpdate` subscription on the `SenderManager` for the `SenderMonitor` to subscribe to.

Reply:
Ah I see a subscription created via `SenderMonitor.SubscribeMaxFloat()` only receives max float updates with the max float for a sender changes when a a ticket is redeemed.

Instead of adding a subscription method to `SenderManager` just so `SenderMonitor` can receive max float updates, what do you think about fetching the max float from `SenderMonitor` using the `MaxFloat()` at a regular interval? This may simplify things and would eliminate the need for the `SenderMonitor.SubscribeMaxFloat()` method as well.

A rough sketch of what this might look like:

```golang
// This is pretty arbitrary, maybe there is a better value
var maxFloatPollInterval = 5 * time.Second

...
ticker := time.NewTicker(maxFloatPollInterval)
for {
    select {
    case <-quit:
        return
    case <-ticker.C:
        maxFloat, err := r.sm.MaxFloat(sender)
        if err != nil {
            // Handle error
        }
        // Check if maxFloat is different from last cached maxFloat
        // If it is different, update the cached maxFloat and send the updated value
        // to listeners
    }
}
```

The downside of this approach is that the server needs to cache maxFloat updates to avoid sending updates to clients even when the maxFloat has not changed. The upside of this approach is that we can avoid managing subscription code in both `SenderMonitor` and `SenderWatcher`.

Reply:
Yeah subscriptions would be more ideal - polling would just accept introducing a chance that updates are not sent exactly when they happen (although this  chance could be minimized by tuning the polling interval) in exchange for reduced code update requirements.

> we'd have to manage a watchdog for either solution

What watchdog are you referring to? The SenderMonitor implementation should already have a watchdog that cleans up sender information stored both on itself as well as on the SenderManager implementation.

> It might be a useful method down the line even while the polling interval would be isolated logic.

Yeah fair enough. If you think that code updates required for adding subscribing to SenderInfo updates on the SenderManager from the SenderMonitor aren't particularly large then I'm cool with taking this approach.

Reply:
I'm mostly just following along on this series of PRs, and don't really have any concrete suggestions right now, but just wanted to point out an area that is still confusing to me from the outside, from this conversation and some others around float monitoring, such as https://github.com/livepeer/go-livepeer/pull/1495#discussion_r432845661 and https://github.com/livepeer/go-livepeer/pull/1513#discussion_r432720399 

This is my current understanding:

(Using the terms 'broadcaster' rather than 'sender' just because it's easier for me to contextualize mentally who is doing what)

* RPC float monitoring does *not* depend on a per-broadcaster subscription. The RPC redeemer blasts out float updates to all orchestrators upon on ticket redemption, regardless of whether the orchestrator is working with the broadcaster. (MonitorMaxFloat)

* Orchestrators still need to create a per-broadcaster subscription for blockchain event monitoring and other bookkeeping. (SubscribeMaxFloat) This is separate from MonitorMaxFloat.

* If a RPC redeemer is running, it does *not* monitor for blockchain events, only winning tickets from other orchestrators. (Or does it?)

My sense has been that these could have been a single call that's abstracted out into a RPC monitor or a local monitor, depending on the mode. Since I think we're planning to build a blockchain monitoring service at some point anyway, this seemed like a sensible first step?

Anyway, this isn't really a request to change anything - just wanted to express (again) a bit of where my confusion was coming from. It would be great to codify the current design under `/doc` so later folks have an easier time in following along, especially because the terms / concepts used in the implementation are a bit on the abstract side.

Reply:
Max float is a composite value of state kept on the `senderMonitor` as well as the `senderWatcher`

`senderWatcher.SubscribeMaxFloatChange` (tentative naming)  would be to tell the `senderMonitor` like "hey the max float for this sender changed , just letting you know in case you wanted to update your package consumer" (which would be true in the case of the `Redeemer`) 

`senderMonitor.SubscribeMaxFloat` would allow consumers of the `senderMonitor` to know when `maxfloat` for a broadcaster changes due to state changes in either `senderWatcher` or `senderMonitor`

Reply:
Actually I spent some more time thinking about this and I don't believe we don't need senderWatcher subscriptions at all. 

So which events would impact a max float change ? 
- Winning ticket redeemed
- Fund Reserve
- Withdraw 
- New Round 

In case of the ticket redemption we already would send a max float update due to the deferred `addFloat` call 

In case of a withdrawal we would have already stopped working with the broadcaster before the deposit and reserve unlocked 

Fund reserve , there seems to be no negative impact here by not immediately sending updates when the reserve gets funded 

New round , would have to update max float values for all senders right now if the pool size changes. I believe this to be a an edge case and impact would be minimal unless there's a huge shift in pool size.

Reply:
> In case of a withdrawal we would have already stopped working with the broadcaster before the deposit and reserve unlocked

Can a winning ticket still draw down locked funds? Eg, a B starts working with an O, puts everything under a withdrawal lock, then suddenly generates a winning ticket. Or is that a non-issue? If so, it seems okay.

Am guessing that the next call to MaxFloat will exclude locked funds from the balance? So the O can decline to work with the B at that point.

Reply:
> Fund reserve , there seems to be no negative impact here by not immediately sending updates when the reserve gets funded unless the segment length would be adequately long to be able to refund a reserve that would have been insufficient by the time the next segment rolls around.

The negative impact is that the last max float that the redeemer sends to O will be inaccurate until a winning ticket from B is redeemed which can affect the ticket face value and price that O advertises to B.

> I believe this to be a an edge case and impact would be minimal unless there's a huge shift in pool size.

While the difference between the perceived max float and the actual max float might not be large when there is a small change in the on-chain pool size, I think we should account for these changes for the reasons mentioned above.

> senderWatcher.SubscribeMaxFloatChange (tentative naming)

`SubscribeReserveUpdate()` might be a more accurate name since the `SenderWatcher` is not aware of the concept of a max float - all it knows is that the on-chain reserve has changed.

Reply:
This has been fixed in #1495